### PR TITLE
Refactor baremetal machine manager tests

### PR DESCRIPTION
**Pull request by maelk**
to use ginkgo and gomega

### CODE REVIEW

File: baremetal/baremetalmachine_manager_test.go [hunk 1]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
+			Expect(bmMachine.Status.ErrorMessage).To(BeNil())
 		},
-		Spec: bmh.BareMetalHostSpec{
-			ConsumerRef: &corev1.ObjectReference{
-				Name:       "someothermachine",
-				Namespace:  "myns",
-				Kind:       "BMMachine",
-				APIVersion: capbm.GroupVersion.String(),
+		Entry("No errors", capbm.BareMetalMachine{}),
+		Entry("Existing error message", capbm.BareMetalMachine{
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: Overwrite Existing error message ?

File: baremetal/baremetalmachine_manager_test.go [hunk 2]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
+				Namespace: "someotherns",
+			},
+		}
+		discoveredHost := bmh.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "discoveredHost",
+				Namespace: "myns",
+			},
+			Status: bmh.BareMetalHostStatus{
+				ErrorMessage: "this host is discovered and not usable",
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: but not ?

File: baremetal/baremetalmachine_manager_test.go [hunk 3]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
-	}
-
-	for name, tc := range testCases {
-		t.Logf("## TC-%s ##", name)
-		c := fakeclient.NewFakeClientWithScheme(setupSchemeMm(), tc.Host)
-
-		bmmconfig, infrastructureRef := newConfig(t, tc.UserDataNamespace, map[string]string{}, []capbm.HostSelectorRequirement{})
-		machine := newMachine("machine1", "", infrastructureRef)
+		}),
+		Entry("Previously provisioned, different image, unchanged",
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: define what resource/item is unchanged.

File: baremetal/baremetalmachine_manager_test.go [hunk 4]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
+			Expected  bool
 		}
 
-		// get the saved host
-		savedHost := bmh.BareMetalHost{}
-		err = c.Get(context.TODO(), client.ObjectKey{Name: tc.Host.Name, Namespace: tc.Host.Namespace}, &savedHost)
-		if err != nil {
-			t.Error(err)
-		}
+		DescribeTable("Test Exists",
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: define what exists

File: baremetal/baremetalmachine_manager_test.go [hunk 5]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
+				Expect(result).To(Equal(tc.Expected))
+			},
+			Entry("Failed to find the existing host", testCaseExists{
+				Machine: &capi.Machine{},
+				BMMachine: newBareMetalMachine("mybmmachine", nil, nil, nil,
+					bmmObjectMetaWithSomeAnnotations(),
+				),
+				Expected: true,
+			}),
+			Entry("Found host even though annotation value incorrect", testCaseExists{
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: value is incorrect

File: baremetal/baremetalmachine_manager_test.go [hunk 6]
```diff
@@ -269,536 +265,490 @@ func TestIsProvisioned(t *testing.T) {
...
+			Entry("Should not find the host, annotation value incorrect",
+				testCaseGetHost{
+					Machine: &capi.Machine{},
+					BMMachine: newBareMetalMachine("mybmmachine", nil, nil, nil,
+						bmmObjectMetaWithInvalidAnnotations(),
+					),
+					ExpectPresent: false,
+				},
+			),
+			Entry("Should not find the host, annotation not present", testCaseGetHost{
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: annotation value not present

File: baremetal/baremetalmachine_manager_test.go [hunk 7]
```diff
@@ -843,872 +795,810 @@ func TestGetSetProviderID(t *testing.T) {
...
-				t.Error(err)
-			}
-		} else {
-			if tc.ExpectError {
-				t.Error("Expected an error")
-			}
+		}),
+	)
+
+	Describe("Test small functions", func() {
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: Test Utility functions ? (or what ever the purpose of the "small" functions is)

File: baremetal/baremetalmachine_manager_test.go [hunk 8]
```diff
@@ -843,872 +795,810 @@ func TestGetSetProviderID(t *testing.T) {
...
-		BMMachine           *capbm.BareMetalMachine
-		ExpectedConsumerRef *corev1.ObjectReference
-		ExpectedResult      error
-	}{
-		"Deprovisioning required": {
-			Host:                newBareMetalHost("myhost", bmhSpec, bmh.StateProvisioned, bmhStatus, false),
-			Machine:             newMachine("mymachine", "", nil),
-			BMMachine:           newBareMetalMachine("mybmmachine", nil, nil, nil, bmmObjectMetaWithValidAnnotations),
+		},
+		Entry("Deprovisioning required", testCaseDelete{
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: change required to needed, much like the following test cases
(unless this is one is different)

File: baremetal/baremetalmachine_manager_test.go [hunk 9]
```diff
@@ -843,872 +795,810 @@ func TestGetSetProviderID(t *testing.T) {
...
+					return
+				} else {
+					Expect(err).To(Succeed())
+				}
 
-		if err != nil {
-			t.Error(err)
-			continue
-		}
+				// get the result
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager_test.go]: // get the node(s)

File: baremetal/baremetalmachine_manager_test.go (hunk 6 above)
Comment (maelk) [reply baremetal/baremetalmachine_manager_test.go]: here the whole annotation is missing

**Pull request merged**